### PR TITLE
disttask: add cleanup routine for framework

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #46258 

Problem Summary:

### What is changed and how it works?
add cleanUpLoop to call cleanUpRoutine to each type of tasks.
We cleanUp routine done, transfer finished tasks to history table.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/dispatcher/dispatcher_manager.go [hunk 1]
```diff
@@ -264,9 +272,72 @@ func (dm *Manager) startDispatcher(task *proto.Task) {
...
+var WaitGCFinished = make(chan struct{})
+
+func (dm *Manager) cleanUpFinishedTasks(tasks []*proto.Task) error {
+	for _, task := range tasks {
+		cleanUpFactory := GetDispatcherCleanUpFactory(task.Type)
+		if cleanUpFactory != nil {
+			cleanUp := cleanUpFactory(dm.ctx, task)
+			err := cleanUp.CleanUp()
+			if err != nil {
+				return err
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher_manager.go:325]: move tasks that have success cleanup to history?

File: disttask/framework/dispatcher/dispatcher_manager.go [hunk 2]
```diff
@@ -264,9 +272,72 @@ func (dm *Manager) startDispatcher(task *proto.Task) {
...
+			}
+			// update task meta.
+			_, err = dm.taskMgr.UpdateGlobalTaskAndAddSubTasks(task, nil, task.State)
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher_manager.go:328]: task meta is not always changed

File: disttask/framework/dispatcher/dispatcher_manager.go [hunk 3]
```diff
@@ -264,9 +272,72 @@ func (dm *Manager) startDispatcher(task *proto.Task) {
...
+}
+
+// WaitGCFinished is used to sync the test.
+var WaitGCFinished = make(chan struct{})
+
+func (dm *Manager) cleanUpFinishedTasks(tasks []*proto.Task) error {
+	for _, task := range tasks {
+		cleanUpFactory := GetDispatcherCleanUpFactory(task.Type)
+		if cleanUpFactory != nil {
+			cleanUp := cleanUpFactory(dm.ctx, task)
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher_manager.go:322]: maybe move those 2 param to `CleanUp()`

File: disttask/framework/dispatcher/interface.go [hunk 4]
```diff
@@ -75,22 +75,59 @@ var dispatcherFactoryMap = struct {
...
-func RegisterDispatcherFactory(taskType string, ctor FactoryFn) {
+func RegisterDispatcherFactory(taskType string, ctor dispatcherFactoryFn) {
 	dispatcherFactoryMap.Lock()
 	defer dispatcherFactoryMap.Unlock()
 	dispatcherFactoryMap.m[taskType] = ctor
 }
 
 // GetDispatcherFactory is used to get the dispatcher factory.
-func GetDispatcherFactory(taskType string) FactoryFn {
+func GetDispatcherFactory(taskType string) dispatcherFactoryFn {
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/interface.go:85]: seems we can make it un-exported, below GetDispatcherCleanUpFactory too

File: disttask/framework/handle/handle.go [hunk 5]
```diff
@@ -79,7 +79,13 @@ func WaitGlobalTask(ctx context.Context, globalTask *proto.Task) error {
 			}
 
 			if found == nil {
-				return errors.Errorf("cannot find global task with ID %d", globalTask.ID)
+				found, err = globalTaskManager.GetTaskByIDFromHistory(globalTask.ID)
```
Comment (D3Hunter) [inline disttask/framework/handle/handle.go:82]: the task don't need to wait framework cleanup

File: disttask/importinto/clean_s3.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	taskMeta := &TaskMeta{}
+	err := json.Unmarshal(c.task.Meta, taskMeta)
+	if err != nil {
+		return err
+	}
+	// Not use cloud storage, no need to cleanUp.
+	if taskMeta.Plan.CloudStorageURI == "" {
+		return nil
+	}
+	defer callLog.End(zap.InfoLevel, nil)
```
Comment (D3Hunter) [inline disttask/importinto/clean_s3.go:61]: should be called after `BeginTask` to pair up

File: disttask/importinto/clean_s3.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+		ctx:  ctx,
+		task: task,
+	}
+}
+
+// CleanUp implements the CleanUpRoutine.CleanUp interface.
+func (c *ImportCleanUpS3) CleanUp() error {
+	// we can only clean up files after all write&ingest subtasks are finished,
+	// since they might share the same file.
+	// TODO: maybe add a way to notify user that there are files left in global sorted storage.
```
Comment (D3Hunter) [inline disttask/importinto/clean_s3.go:49]: can be removed

File: disttask/importinto/clean_s3.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+		logger.Warn("failed to init data store", zap.Error(err))
+		return err
+	}
+	if err = external.CleanUpFiles(c.ctx, controller.GlobalSortStore,
+		strconv.Itoa(int(c.task.ID))); err != nil {
+		logger.Warn("failed to clean up files of task", zap.Error(err))
+		return err
+	}
+	// Only redact sensitive info after cleanUpFiles success.
+	redactSensitiveInfo(c.task, taskMeta)
```
Comment (D3Hunter) [inline disttask/importinto/clean_s3.go:78]: need call this even we're not using global sort

File: tests/realtikvtest/importintotest/import_into_test.go [hunk 9]
```diff
@@ -779,6 +779,10 @@ func (s *mockGCSSuite) checkTaskMetaRedacted(jobID int64) {
 	s.NoError(err)
 	globalTask, err2 := globalTaskManager.GetGlobalTaskByKey(taskKey)
 	s.NoError(err2)
+	if globalTask == nil {
+		globalTask, err2 = globalTaskManager.GetGlobalTaskByKeyFromHistory(taskKey)
```
Comment (D3Hunter) [inline tests/realtikvtest/importintotest/import_into_test.go:783]: add a method to query both table?

File: disttask/framework/storage/task_table.go [hunk 10]
```diff
@@ -860,6 +903,57 @@ func (stm *TaskManager) GC() error {
...
 }
 
+// TransferTasks2History transfer the selected tasks into tidb_global_task_history table by taskIDs.
+func (stm *TaskManager) TransferTasks2History(tasks []*proto.Task) error {
+	if len(tasks) == 0 {
+		return nil
+	}
+	return stm.WithNewTxn(stm.ctx, func(se sessionctx.Context) error {
+		insertSQL := new(strings.Builder)
+		if err := sqlexec.FormatSQL(insertSQL, "insert into mysql.tidb_global_task_history select * from mysql.tidb_global_task where id in ("); err != nil {
```
Comment (D3Hunter) [inline disttask/framework/storage/task_table.go:913]: so ugly to use `FormatSQL`, generate slice of `%?` using `strings.Join`?

File: tests/realtikvtest/importintotest/import_into_test.go (hunk 9 above)
Comment (ywqzzy) [reply tests/realtikvtest/importintotest/import_into_test.go:783]: > add a method to query both table?

Ok

**Pull request merged**